chain/chain/src/doomslug.rs

```diff
@@ -0,0 +1,1004 @@
+use near_crypto::Signer;
+use near_primitives::block::Approval;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{AccountId, Balance, BlockHeight, BlockHeightDelta, ValidatorStake};
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+/// Have that many iterations in the timer instead of `loop` to prevent potential bugs from blocking
+/// the node
+const MAX_TIMER_ITERS: usize = 20;
+
+/// How many heights ahead to track approvals. This needs to be sufficiently large so that we can
+/// recover after rather long network interruption, but not too large to consume too much memory if
+/// someone in the network spams with invalid approvals. Note that we will only store approvals for
+/// heights that are targeting us, which is once per as many heights as there are block producers,
+/// thus 10_000 heights in practice will mean on the order of one hundred entries.
+const MAX_HEIGHTS_AHEAD_TO_STORE_APPROVALS: BlockHeight = 10_000;
+
+/// The threshold for doomslug to create a block.
+/// `HalfStake` means the block can only be produced if at least half of the stake is approvign it,
+///             and is what should be used in production (and what guarantees doomslug finality)
+/// `NoApprovals` means the block production is not blocked on approvals. This is used
+///             in many tests (e.g. `cross_shard_tx`) to create lots of forkfulness.
+#[derive(PartialEq, Eq, Debug, Clone, Copy)]
+pub enum DoomslugThresholdMode {
+    NoApprovals,
+    HalfStake,
+}
+
+/// The result of processing an approval.
+///
+/// `None`
+/// `PassedThreshold(when)` - after processing this approval the block has passed the threshold set by
```

Comment:
I'd suggest to move comments per item instead of having them as a top-level for enum.